src/index/utxoscriptindex.cpp

```diff
@@ -0,0 +1,455 @@
+    for (unsigned int i = 0; i < tx.vout.size(); i++)
+    {
+        if (tx.vout[i].IsNull() || tx.vout[i].scriptPubKey.IsUnspendable())
+            continue;
+
+        auto utxoKey = ScriptIndexHash(tx.vout[i].scriptPubKey);
+        utxoCacheMempool[utxoKey].insert(COutPoint(tx.GetHash(), static_cast<uint32_t>(i)));
+    }
+}
+
+void UtxoScriptIndex::TransactionRemovedFromMempool(const CTransactionRef &ptx) {
+    LOCK(cs_utxoCacheMempool);
+    const CTransaction &tx = *ptx;
+    for (unsigned int i = 0; i < tx.vout.size(); i++)
+    {
+        if (tx.vout[i].IsNull() || tx.vout[i].scriptPubKey.IsUnspendable())
+            continue;
+
+        std::map<ScriptHash, SerializableUtxoSet>::iterator it = utxoCacheMempool.find(ScriptIndexHash(tx.vout[i].scriptPubKey));
+
+        if (it != utxoCacheMempool.end())
+        {
+            it->second.erase(COutPoint(tx.GetHash(), static_cast<uint32_t>(i)));
+            if (it->second.empty())
+                utxoCacheMempool.erase(it);
+        }
+    }
+}
+
+void UtxoScriptIndex::BlockConnected(const std::shared_ptr<const CBlock> &block,
+                                     const CBlockIndex *pindex,
+                                     const std::shared_ptr<const CBlockUndo> &blockundo,
+                                     const std::vector<CTransactionRef> &txnConflicted)
+{
+    {
+        LOCK(cs_utxoCache);
+        removeSpentUtxosOnTipConnected(*block, *blockundo);
+        addNewUtxosOnTipConnected(*block);
+    }
+    WriteBestBlock(block->GetHash());
+}
+
+void UtxoScriptIndex::BlockDisconnected(const std::shared_ptr<const CBlock> &block,
+                                        const std::shared_ptr<const CBlockUndo> &blockundo) {
+    assert(block->vtx.size() > 0);
+    LOCK(cs_utxoCache);
+    restoreSpentUtxosOnTipDisconnected(*block, *blockundo);
+    removeUtxosOnTipDisconnected(*block);
+}
+
+BaseIndex::DB& UtxoScriptIndex::GetDB() const { return *m_db; }
+
+void UtxoScriptIndex::removeUtxo(const CTxOut& txout, const COutPoint outpoint)
+{
+    if(txout.IsNull() || txout.scriptPubKey.IsUnspendable())
+        return;
+    const ScriptHash utxoCacheKey = ScriptIndexHash(txout.scriptPubKey);
+    loadToCache(utxoCacheKey);
+    if(utxoCache.count(utxoCacheKey) == 0)
+        return;
+    SerializableUtxoSet& utxoCacheSet = utxoCache.at(utxoCacheKey);
+        utxoCacheSet.erase(outpoint);
+}
+
+void UtxoScriptIndex::removeSpentUtxosOnTipConnected(const CBlock& block, const CBlockUndo& blockundo)
+{
+
+    for(unsigned int i = 1; i < block.vtx.size(); ++i)
+    {
+        for(unsigned int j = 0; j < (block.vtx[i])->vin.size(); ++j)
+        {
+            removeUtxo(blockundo.vtxundo[i-1].vprevout[j].out,
+                       block.vtx[i]->vin[j].prevout);
+        }
+    }
+}
+
+void UtxoScriptIndex::removeUtxosOnTipDisconnected(const CBlock& block)
+{
+    if(block.vtx.size() == 0)
+        return;
+
+    unsigned int i = block.vtx.size() - 1;
+    while(true)
```

Comment:
This could be a `do{ } while(i != 0)`

Reply:
Loop needs to execute for `i==0` too (that is block's tx with index 0); `i` is unsigned so after decrementation it overflows and such loop would be indefinite